File: SkyBridge Compass iOS/SkyBridgeCompassiOS/Sources/Core/RemoteConnection/RemoteDesktop/RemoteControlSecureEnvelope.swift
import CryptoKit
import Foundation

enum RemoteControlSecurePacketType: UInt8, Sendable, Hashable {
    case control = 1
    case screen = 2
    case audio = 3
}

struct RemoteControlSecureOpenedPayload: Sendable {
    let packetType: RemoteControlSecurePacketType
    let direction: UInt8
    let sessionHash: UInt64
    let transcriptPrefix: UInt64
    let epoch: UInt32
    let counter: UInt64
    let payload: Data
}

enum RemoteControlSecureReplayRejectionReason: String, Sendable {
    case duplicateCounter = "duplicate-counter"
    case counterOutsideWindow = "counter-outside-window"
}

enum RemoteControlSecureEnvelopeError: Error, LocalizedError, Equatable {
    case malformed
    case unsupportedMagic(UInt32)
    case unsupportedVersion(UInt8)
    case unsupportedPacketType(UInt8)
    case packetTypeMismatch(expected: [RemoteControlSecurePacketType], actual: RemoteControlSecurePacketType)
    case directionMismatch(expected: UInt8, actual: UInt8)
    case sessionMismatch(expected: UInt64, actual: UInt64)
    case transcriptMismatch(expected: UInt64, actual: UInt64)
    case epochMismatch(expected: UInt32, actual: UInt32)
    case authenticationFailed(packetType: RemoteControlSecurePacketType, counter: UInt64)
    case invalidCounter(UInt64)
    case replayDetected(
        packetType: RemoteControlSecurePacketType,
        counter: UInt64,
        highestCounter: UInt64,
        reason: RemoteControlSecureReplayRejectionReason
    )

    var errorDescription: String? {
        switch self {
        case .malformed:
            return "malformed secure envelope"
        case .unsupportedMagic(let magic):
            return "unsupported secure envelope magic=\(magic)"
        case .unsupportedVersion(let version):
            return "unsupported secure envelope version=\(version)"
        case .unsupportedPacketType(let raw):
            return "unsupported secure envelope packetType=\(raw)"
        case .packetTypeMismatch(let expected, let actual):
            let expectedRaw = expected.map { String($0.rawValue) }.joined(separator: ",")
            return "secure envelope packetType mismatch expected=\(expectedRaw) actual=\(actual.rawValue)"
        case .directionMismatch(let expected, let actual):
            return "secure envelope direction mismatch expected=\(expected) actual=\(actual)"
        case .sessionMismatch(let expected, let actual):
            return "secure envelope session mismatch expected=\(expected) actual=\(actual)"
        case .transcriptMismatch(let expected, let actual):
            return "secure envelope transcript mismatch expected=\(expected) actual=\(actual)"
        case .epochMismatch(let expected, let actual):
            return "secure envelope epoch mismatch expected=\(expected) actual=\(actual)"
        case .authenticationFailed(let packetType, let counter):
            return "secure envelope authentication failed packetType=\(packetType.rawValue) counter=\(counter)"
        case .invalidCounter(let counter):
            return "secure envelope invalid counter=\(counter)"
        case .replayDetected(let packetType, let counter, let highestCounter, let reason):
            return "secure envelope replay detected packetType=\(packetType.rawValue) counter=\(counter) highestCounter=\(highestCounter) reason=\(reason.rawValue)"
        }
    }
}

struct RemoteControlSecureReplayWindow: Sendable {
    private struct ReplayScope: Hashable, Sendable {
        let packetType: RemoteControlSecurePacketType
        let direction: UInt8
        let sessionHash: UInt64
        let transcriptPrefix: UInt64
        let epoch: UInt32
    }

    private struct ReplayLane: Sendable {
        var highestCounter: UInt64 = 0
        var recordedCounters: Set<UInt64> = []
    }

    private static let windowSize: UInt64 = 1024
    private var lanes: [ReplayScope: ReplayLane] = [:]

    mutating func validateAndRecord(_ openedPayload: RemoteControlSecureOpenedPayload) throws {
        guard openedPayload.counter > 0 else {
            throw RemoteControlSecureEnvelopeError.invalidCounter(openedPayload.counter)
        }

        let scope = ReplayScope(
            packetType: openedPayload.packetType,
            direction: openedPayload.direction,
            sessionHash: openedPayload.sessionHash,
            transcriptPrefix: openedPayload.transcriptPrefix,
            epoch: openedPayload.epoch
        )
        var lane = lanes[scope] ?? ReplayLane()
        let highestCounter = lane.highestCounter

        if openedPayload.counter > highestCounter {
            lane.highestCounter = openedPayload.counter
            lane.recordedCounters.insert(openedPayload.counter)
            pruneRecordedCounters(in: &lane)
            lanes[scope] = lane
            return
        }

        let counterDistance = highestCounter - openedPayload.counter
        guard counterDistance < Self.windowSize else {
            throw RemoteControlSecureEnvelopeError.replayDetected(
                packetType: openedPayload.packetType,
                counter: openedPayload.counter,
                highestCounter: highestCounter,
                reason: .counterOutsideWindow
            )
        }
        guard !lane.recordedCounters.contains(openedPayload.counter) else {
            throw RemoteControlSecureEnvelopeError.replayDetected(
                packetType: openedPayload.packetType,
                counter: openedPayload.counter,
                highestCounter: highestCounter,
                reason: .duplicateCounter
            )
        }

        lane.recordedCounters.insert(openedPayload.counter)
        lanes[scope] = lane
    }

    private func pruneRecordedCounters(in lane: inout ReplayLane) {
        let minimumCounterToKeep: UInt64
        if lane.highestCounter > Self.windowSize {
            minimumCounterToKeep = lane.highestCounter - Self.windowSize + 1
        } else {
            minimumCounterToKeep = 1
        }
        lane.recordedCounters = lane.recordedCounters.filter { $0 >= minimumCounterToKeep }
    }
}

enum RemoteControlSecureEnvelope {
    static let overheadBytes = headerLength + tagLength

    private static let magic: UInt32 = 0x5342_5243 // "SBRC"
    private static let version: UInt8 = 1
    private static let headerLength = 52
    private static let tagLength = 16
    private static let epoch: UInt32 = 0
    private static let directionInitiatorToResponder: UInt8 = 1
    private static let directionResponderToInitiator: UInt8 = 2

    static func seal(
        _ plaintext: Data,
        keys: SessionKeys,
        packetType: RemoteControlSecurePacketType,
        counter: UInt64
    ) throws -> Data {
        guard counter > 0 else {
            throw RemoteControlSecureEnvelopeError.invalidCounter(counter)
        }
        let nonce = AES.GCM.Nonce()
        let nonceData = nonce.withUnsafeBytes { Data($0) }
        var header = Data()
        header.reserveCapacity(headerLength)
        appendUInt32(magic, to: &header)
        header.append(version)
        header.append(UInt8(headerLength))
        header.append(packetType.rawValue)
        header.append(sendDirection(for: keys))
        appendUInt64(sessionIdHash(keys.sessionId), to: &header)
        appendUInt64(transcriptPrefix(keys.transcriptHash), to: &header)
        appendUInt32(epoch, to: &header)
        appendUInt64(counter, to: &header)
        appendUInt32(UInt32(plaintext.count), to: &header)
        header.append(nonceData)
        precondition(header.count == headerLength)

        let sealed = try AES.GCM.seal(
            plaintext,
            using: SymmetricKey(data: keys.sendKey),
            nonce: nonce,
            authenticating: header
        )
        var output = header
        output.append(sealed.ciphertext)
        output.append(sealed.tag)
        return output
    }

    static func open(
        _ packet: Data,
        keys: SessionKeys,
        allowedPacketTypes: Set<RemoteControlSecurePacketType>
    ) throws -> RemoteControlSecureOpenedPayload {
        let parsed = try parseHeader(packet)
        guard allowedPacketTypes.contains(parsed.packetType) else {
            throw RemoteControlSecureEnvelopeError.packetTypeMismatch(
                expected: Array(allowedPacketTypes).sorted { $0.rawValue < $1.rawValue },
                actual: parsed.packetType
            )
        }
        let expectedDirection = receiveDirection(for: keys)
        guard parsed.direction == expectedDirection else {
            throw RemoteControlSecureEnvelopeError.directionMismatch(expected: expectedDirection, actual: parsed.direction)
        }
        let expectedSessionHash = sessionIdHash(keys.sessionId)
        guard parsed.sessionHash == expectedSessionHash else {
            throw RemoteControlSecureEnvelopeError.sessionMismatch(expected: expectedSessionHash, actual: parsed.sessionHash)
        }
        let expectedTranscriptPrefix = transcriptPrefix(keys.transcriptHash)
        guard parsed.transcriptPrefix == expectedTranscriptPrefix else {
            throw RemoteControlSecureEnvelopeError.transcriptMismatch(
                expected: expectedTranscriptPrefix,
                actual: parsed.transcriptPrefix
            )
        }
        guard parsed.epoch == epoch else {
            throw RemoteControlSecureEnvelopeError.epochMismatch(expected: epoch, actual: parsed.epoch)
        }
        guard parsed.counter > 0 else {
            throw RemoteControlSecureEnvelopeError.invalidCounter(parsed.counter)
        }

        let ciphertextStart = packet.startIndex + headerLength
        let ciphertextEnd = ciphertextStart + parsed.payloadLength
        let ciphertext = packet[ciphertextStart..<ciphertextEnd]
        let tag = packet[packet.index(packet.endIndex, offsetBy: -tagLength)..<packet.endIndex]
        let nonce = try AES.GCM.Nonce(data: parsed.nonce)
        let box = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tag)
        do {
            let payload = try AES.GCM.open(
                box,
                using: SymmetricKey(data: keys.receiveKey),
                authenticating: packet.prefix(headerLength)
            )
            return RemoteControlSecureOpenedPayload(
                packetType: parsed.packetType,
                direction: parsed.direction,
                sessionHash: parsed.sessionHash,
                transcriptPrefix: parsed.transcriptPrefix,
                epoch: parsed.epoch,
                counter: parsed.counter,
                payload: payload
            )
        } catch {
            throw RemoteControlSecureEnvelopeError.authenticationFailed(
                packetType: parsed.packetType,
                counter: parsed.counter
            )
        }
    }

    private struct ParsedHeader {
        let packetType: RemoteControlSecurePacketType
        let direction: UInt8
        let sessionHash: UInt64
        let transcriptPrefix: UInt64
        let epoch: UInt32
        let counter: UInt64
        let payloadLength: Int
        let nonce: Data
    }

    private static func parseHeader(_ packet: Data) throws -> ParsedHeader {
        guard packet.count >= headerLength + tagLength else {
            throw RemoteControlSecureEnvelopeError.malformed
        }
        let magic = readUInt32(packet, at: 0)
        guard magic == self.magic else {
            throw RemoteControlSecureEnvelopeError.unsupportedMagic(magic)
        }
        let version = packet[packet.startIndex + 4]
        guard version == self.version else {
            throw RemoteControlSecureEnvelopeError.unsupportedVersion(version)
        }
        let encodedHeaderLength = Int(packet[packet.startIndex + 5])
        guard encodedHeaderLength == headerLength else {
            throw RemoteControlSecureEnvelopeError.malformed
        }
        let packetTypeRaw = packet[packet.startIndex + 6]
        guard let packetType = RemoteControlSecurePacketType(rawValue: packetTypeRaw) else {
            throw RemoteControlSecureEnvelopeError.unsupportedPacketType(packetTypeRaw)
        }
        let payloadLength = Int(readUInt32(packet, at: 36))
        guard packet.count == headerLength + payloadLength + tagLength else {
            throw RemoteControlSecureEnvelopeError.malformed
        }
        return ParsedHeader(
            packetType: packetType,
            direction: packet[packet.startIndex + 7],
            sessionHash: readUInt64(packet, at: 8),
            transcriptPrefix: readUInt64(packet, at: 16),
            epoch: readUInt32(packet, at: 24),
            counter: readUInt64(packet, at: 28),
            payloadLength: payloadLength,
            nonce: Data(packet[(packet.startIndex + 40)..<(packet.startIndex + 52)])
        )
    }

    private static func sendDirection(for keys: SessionKeys) -> UInt8 {
        keys.role == .initiator ? directionInitiatorToResponder : directionResponderToInitiator
    }

    private static func receiveDirection(for keys: SessionKeys) -> UInt8 {
        keys.role == .initiator ? directionResponderToInitiator : directionInitiatorToResponder
    }

    private static func sessionIdHash(_ sessionId: String) -> UInt64 {
        var input = Data("SkyBridge-RemoteControl-Session-v1|".utf8)
        input.append(Data(sessionId.utf8))
        return firstUInt64(of: SHA256.hash(data: input))
    }

    private static func transcriptPrefix(_ transcriptHash: Data) -> UInt64 {
        var input = Data("SkyBridge-RemoteControl-Transcript-v1|".utf8)
        input.append(transcriptHash)
        return firstUInt64(of: SHA256.hash(data: input))
    }

    private static func firstUInt64<D: Sequence>(of digest: D) -> UInt64 where D.Element == UInt8 {
        digest.prefix(8).reduce(UInt64(0)) { partial, byte in
            (partial << 8) | UInt64(byte)
        }
    }

    private static func appendUInt32(_ value: UInt32, to data: inout Data) {
        data.append(UInt8((value >> 24) & 0xff))
        data.append(UInt8((value >> 16) & 0xff))
        data.append(UInt8((value >> 8) & 0xff))
        data.append(UInt8(value & 0xff))
    }

    private static func appendUInt64(_ value: UInt64, to data: inout Data) {
        for shift in stride(from: 56, through: 0, by: -8) {
            data.append(UInt8((value >> UInt64(shift)) & 0xff))
        }
    }

    private static func readUInt32(_ data: Data, at offset: Int) -> UInt32 {
        (UInt32(data[data.startIndex + offset]) << 24)
            | (UInt32(data[data.startIndex + offset + 1]) << 16)
            | (UInt32(data[data.startIndex + offset + 2]) << 8)
            | UInt32(data[data.startIndex + offset + 3])
    }

    private static func readUInt64(_ data: Data, at offset: Int) -> UInt64 {
        var value: UInt64 = 0
        for index in 0..<8 {
            value = (value << 8) | UInt64(data[data.startIndex + offset + index])
        }
        return value
    }
}

File: platforms/android/core/src/main/kotlin/com/skybridge/compass/remotecontrol/secure/RemoteControlSecureEnvelope.kt
package com.skybridge.compass.remotecontrol.secure

import com.skybridge.compass.shared.p2p.P2PSessionIds
import java.security.MessageDigest
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * SBRC ("SkyBridge-RemoteControl") secure envelope for LAN remote-control frames.
 *
 * This is the Android mirror of Apple
 * `Sources/SkyBridgeCore/RemoteControl/RemoteControlSecureEnvelope.swift`.
 */
object RemoteControlSecureEnvelope {
    enum class PacketType(val raw: Int) {
        CONTROL(1),
        SCREEN(2),
        AUDIO(3);

        companion object {
            fun fromRaw(raw: Int): PacketType? = entries.firstOrNull { it.raw == raw }
        }
    }

    enum class Role { INITIATOR, RESPONDER }

    data class Opened(
        val packetType: PacketType,
        val direction: Int,
        val sessionHash: Long,
        val transcriptPrefix: Long,
        val epoch: Long,
        val counter: Long,
        val payload: ByteArray
    )

    sealed class EnvelopeException(message: String) : Exception(message) {
        class Malformed : EnvelopeException("malformed remote-control secure envelope")
        class UnsupportedMagic(val magic: Long) :
            EnvelopeException("unsupported remote-control secure envelope magic=$magic")
        class UnsupportedVersion(val version: Int) :
            EnvelopeException("unsupported remote-control secure envelope version=$version")
        class UnsupportedPacketType(val raw: Int) :
            EnvelopeException("unsupported remote-control secure envelope packetType=$raw")
        class PacketTypeMismatch(val expected: Set<PacketType>, val actual: PacketType) :
            EnvelopeException(
                "remote-control secure envelope packetType mismatch expected=" +
                    expected.map { it.raw }.sorted().joinToString(",") +
                    " actual=${actual.raw}"
            )
        class DirectionMismatch(val expected: Int, val actual: Int) :
            EnvelopeException("remote-control secure envelope direction mismatch expected=$expected actual=$actual")
        class SessionMismatch(val expected: Long, val actual: Long) :
            EnvelopeException("remote-control secure envelope session mismatch expected=$expected actual=$actual")
        class TranscriptMismatch(val expected: Long, val actual: Long) :
            EnvelopeException("remote-control secure envelope transcript mismatch expected=$expected actual=$actual")
        class EpochMismatch(val expected: Long, val actual: Long) :
            EnvelopeException("remote-control secure envelope epoch mismatch expected=$expected actual=$actual")
        class AuthenticationFailed(val packetType: PacketType, val counter: Long) :
            EnvelopeException(
                "remote-control secure envelope authentication failed packetType=${packetType.raw} counter=$counter"
            )
        class InvalidCounter(val counter: Long) :
            EnvelopeException("remote-control secure envelope invalid counter=$counter")
        class ReplayDetected(
            val packetType: PacketType,
            val counter: Long,
            val highestCounter: Long,
            val reason: String
        ) : EnvelopeException(
            "remote-control secure envelope replay detected packetType=${packetType.raw} " +
                "counter=$counter highestCounter=$highestCounter reason=$reason"
        )
    }

    private const val MAGIC: Long = 0x5342_5243L // "SBRC"
    private const val VERSION: Int = 1
    private const val HEADER_LENGTH: Int = 52
    private const val TAG_LENGTH: Int = 16
    private const val NONCE_LENGTH: Int = 12
    private const val EPOCH: Long = 0L
    private const val DIRECTION_INITIATOR_TO_RESPONDER: Int = 1
    private const val DIRECTION_RESPONDER_TO_INITIATOR: Int = 2
    private const val GCM_TAG_BITS: Int = 128

    const val OVERHEAD_BYTES: Int = HEADER_LENGTH + TAG_LENGTH

    private val rng = SecureRandom()

    fun deterministicSessionId(transcriptHash: ByteArray): String =
        P2PSessionIds.deterministicSessionId(transcriptHash)

    fun seal(
        plaintext: ByteArray,
        sendKey: ByteArray,
        role: Role,
        sessionId: String,
        transcriptHash: ByteArray,
        packetType: PacketType,
        counter: Long
    ): ByteArray {
        require(sendKey.size == 32) { "AES-256 key must be 32 bytes" }
        if (counter <= 0L) throw EnvelopeException.InvalidCounter(counter)
        if (plaintext.size.toLong() > 0xFFFF_FFFFL) throw EnvelopeException.Malformed()

        val nonce = ByteArray(NONCE_LENGTH).also { rng.nextBytes(it) }
        val header = ByteArray(HEADER_LENGTH)
        putUInt32(header, 0, MAGIC)
        header[4] = VERSION.toByte()
        header[5] = HEADER_LENGTH.toByte()
        header[6] = packetType.raw.toByte()
        header[7] = sendDirection(role).toByte()
        putUInt64(header, 8, sessionIdHash(sessionId))
        putUInt64(header, 16, transcriptPrefix(transcriptHash))
        putUInt32(header, 24, EPOCH)
        putUInt64(header, 28, counter)
        putUInt32(header, 36, plaintext.size.toLong())
        System.arraycopy(nonce, 0, header, 40, NONCE_LENGTH)

        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(sendKey, "AES"), GCMParameterSpec(GCM_TAG_BITS, nonce))
        cipher.updateAAD(header)
        val ciphertextAndTag = cipher.doFinal(plaintext)

        return header + ciphertextAndTag
    }

    fun open(
        packet: ByteArray,
        receiveKey: ByteArray,
        role: Role,
        sessionId: String,
        transcriptHash: ByteArray,
        allowedPacketTypes: Set<PacketType>
    ): Opened {
        require(receiveKey.size == 32) { "AES-256 key must be 32 bytes" }
        val parsed = parseHeader(packet)

        if (!allowedPacketTypes.contains(parsed.packetType)) {
            throw EnvelopeException.PacketTypeMismatch(allowedPacketTypes, parsed.packetType)
        }
        val expectedDirection = receiveDirection(role)
        if (parsed.direction != expectedDirection) {
            throw EnvelopeException.DirectionMismatch(expectedDirection, parsed.direction)
        }
        val expectedSessionHash = sessionIdHash(sessionId)
        if (parsed.sessionHash != expectedSessionHash) {
            throw EnvelopeException.SessionMismatch(expectedSessionHash, parsed.sessionHash)
        }
        val expectedTranscriptPrefix = transcriptPrefix(transcriptHash)
        if (parsed.transcriptPrefix != expectedTranscriptPrefix) {
            throw EnvelopeException.TranscriptMismatch(expectedTranscriptPrefix, parsed.transcriptPrefix)
        }
        if (parsed.epoch != EPOCH) {
            throw EnvelopeException.EpochMismatch(EPOCH, parsed.epoch)
        }
        if (parsed.counter <= 0L) {
            throw EnvelopeException.InvalidCounter(parsed.counter)
        }

        val ciphertextStart = HEADER_LENGTH
        val ciphertextEnd = ciphertextStart + parsed.payloadLength
        val nonce = packet.copyOfRange(40, 52)
        val ciphertextAndTag = packet.copyOfRange(ciphertextStart, ciphertextEnd + TAG_LENGTH)
        val payload = try {
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(receiveKey, "AES"), GCMParameterSpec(GCM_TAG_BITS, nonce))
            cipher.updateAAD(packet.copyOfRange(0, HEADER_LENGTH))
            cipher.doFinal(ciphertextAndTag)
        } catch (t: Throwable) {
            throw EnvelopeException.AuthenticationFailed(parsed.packetType, parsed.counter)
        }

        return Opened(
            packetType = parsed.packetType,
            direction = parsed.direction,
            sessionHash = parsed.sessionHash,
            transcriptPrefix = parsed.transcriptPrefix,
            epoch = parsed.epoch,
            counter = parsed.counter,
            payload = payload
        )
    }

    private data class ParsedHeader(
        val packetType: PacketType,
        val direction: Int,
        val sessionHash: Long,
        val transcriptPrefix: Long,
        val epoch: Long,
        val counter: Long,
        val payloadLength: Int
    )

    private fun parseHeader(packet: ByteArray): ParsedHeader {
        if (packet.size < HEADER_LENGTH + TAG_LENGTH) throw EnvelopeException.Malformed()

        val magic = readUInt32(packet, 0)
        if (magic != MAGIC) throw EnvelopeException.UnsupportedMagic(magic)

        val version = packet[4].toInt() and 0xFF
        if (version != VERSION) throw EnvelopeException.UnsupportedVersion(version)

        val encodedHeaderLength = packet[5].toInt() and 0xFF
        if (encodedHeaderLength != HEADER_LENGTH) throw EnvelopeException.Malformed()

        val packetTypeRaw = packet[6].toInt() and 0xFF
        val packetType = PacketType.fromRaw(packetTypeRaw)
            ?: throw EnvelopeException.UnsupportedPacketType(packetTypeRaw)

        val payloadLength = readUInt32(packet, 36)
        if (packet.size.toLong() != HEADER_LENGTH.toLong() + payloadLength + TAG_LENGTH.toLong()) {
            throw EnvelopeException.Malformed()
        }

        return ParsedHeader(
            packetType = packetType,
            direction = packet[7].toInt() and 0xFF,
            sessionHash = readUInt64(packet, 8),
            transcriptPrefix = readUInt64(packet, 16),
            epoch = readUInt32(packet, 24),
            counter = readUInt64(packet, 28),
            payloadLength = payloadLength.toInt()
        )
    }

    class ReplayWindow {
        private data class Scope(
            val packetType: PacketType,
            val direction: Int,
            val sessionHash: Long,
            val transcriptPrefix: Long,
            val epoch: Long
        )

        private class Lane {
            var highestCounter: Long = 0L
            val recordedCounters: MutableSet<Long> = HashSet()
        }

        private val lanes = HashMap<Scope, Lane>()

        fun validateAndRecord(opened: Opened) {
            if (opened.counter <= 0L) throw EnvelopeException.InvalidCounter(opened.counter)

            val scope = Scope(
                packetType = opened.packetType,
                direction = opened.direction,
                sessionHash = opened.sessionHash,
                transcriptPrefix = opened.transcriptPrefix,
                epoch = opened.epoch
            )
            val lane = lanes.getOrPut(scope) { Lane() }
            val highest = lane.highestCounter

            if (opened.counter > highest) {
                lane.highestCounter = opened.counter
                lane.recordedCounters.add(opened.counter)
                prune(lane)
                return
            }

            val distance = highest - opened.counter
            if (distance >= WINDOW_SIZE) {
                throw EnvelopeException.ReplayDetected(
                    opened.packetType,
                    opened.counter,
                    highest,
                    "counter-outside-window"
                )
            }
            if (lane.recordedCounters.contains(opened.counter)) {
                throw EnvelopeException.ReplayDetected(
                    opened.packetType,
                    opened.counter,
                    highest,
                    "duplicate-counter"
                )
            }
            lane.recordedCounters.add(opened.counter)
        }

        private fun prune(lane: Lane) {
            val minimumCounterToKeep = if (lane.highestCounter > WINDOW_SIZE) {
                lane.highestCounter - WINDOW_SIZE + 1
            } else {
                1L
            }
            lane.recordedCounters.removeAll { it < minimumCounterToKeep }
        }

        private companion object {
            private const val WINDOW_SIZE: Long = 1024L
        }
    }

    private fun sendDirection(role: Role): Int =
        if (role == Role.INITIATOR) DIRECTION_INITIATOR_TO_RESPONDER else DIRECTION_RESPONDER_TO_INITIATOR

    private fun receiveDirection(role: Role): Int =
        if (role == Role.INITIATOR) DIRECTION_RESPONDER_TO_INITIATOR else DIRECTION_INITIATOR_TO_RESPONDER

    private fun sessionIdHash(sessionId: String): Long {
        val input = "SkyBridge-RemoteControl-Session-v1|".toByteArray(Charsets.UTF_8) +
            sessionId.toByteArray(Charsets.UTF_8)
        return firstUInt64(sha256(input))
    }

    private fun transcriptPrefix(transcriptHash: ByteArray): Long {
        val input = "SkyBridge-RemoteControl-Transcript-v1|".toByteArray(Charsets.UTF_8) + transcriptHash
        return firstUInt64(sha256(input))
    }

    private fun firstUInt64(digest: ByteArray): Long {
        var value = 0L
        for (i in 0 until 8) {
            value = (value shl 8) or (digest[i].toLong() and 0xFF)
        }
        return value
    }

    private fun sha256(data: ByteArray): ByteArray =
        MessageDigest.getInstance("SHA-256").digest(data)

    private fun putUInt32(out: ByteArray, offset: Int, value: Long) {
        out[offset] = ((value ushr 24) and 0xFF).toByte()
        out[offset + 1] = ((value ushr 16) and 0xFF).toByte()
        out[offset + 2] = ((value ushr 8) and 0xFF).toByte()
        out[offset + 3] = (value and 0xFF).toByte()
    }

    private fun putUInt64(out: ByteArray, offset: Int, value: Long) {
        var shift = 56
        var index = 0
        while (shift >= 0) {
            out[offset + index] = ((value ushr shift) and 0xFF).toByte()
            shift -= 8
            index += 1
        }
    }

    private fun readUInt32(data: ByteArray, offset: Int): Long {
        return ((data[offset].toLong() and 0xFF) shl 24) or
            ((data[offset + 1].toLong() and 0xFF) shl 16) or
            ((data[offset + 2].toLong() and 0xFF) shl 8) or
            (data[offset + 3].toLong() and 0xFF)
    }

    private fun readUInt64(data: ByteArray, offset: Int): Long {
        var value = 0L
        for (i in 0 until 8) {
            value = (value shl 8) or (data[offset + i].toLong() and 0xFF)
        }
        return value
    }
}
